From 2eb9f2c0850c0e92877b0e8db7e8173bea5f584e Mon Sep 17 00:00:00 2001
From: Zhuocheng Ding <zhuocheng.ding@intel.com>
Date: Fri, 11 Sep 2020 09:14:30 +0800
Subject: [PATCH] mc146818rtc: add RTC_ALARM QMP event

Emitted whenever the RTC alarm expire time changes.

Tracked-On: OAM-92864
Signed-off-by: Zhuocheng Ding <zhuocheng.ding@intel.com>
---
 hw/rtc/mc146818rtc.c  |  7 ++++++-
 qapi/misc-target.json | 18 ++++++++++++++++++
 2 files changed, 24 insertions(+), 1 deletion(-)

diff --git a/hw/rtc/mc146818rtc.c b/hw/rtc/mc146818rtc.c
index 74ae74bc5c..4db9e946d1 100644
--- a/hw/rtc/mc146818rtc.c
+++ b/hw/rtc/mc146818rtc.c
@@ -269,6 +269,7 @@ static void check_update_timer(RTCState *s)
 {
     uint64_t next_update_time;
     uint64_t guest_nsec;
+    uint64_t next_alarm_time;
     int next_alarm_sec;

     /* From the data sheet: "Holding the dividers in reset prevents
@@ -289,8 +290,12 @@ static void check_update_timer(RTCState *s)
      * for in next_update_time.
      */
     next_alarm_sec = get_next_alarm(s);
-    s->next_alarm_time = next_update_time +
+    next_alarm_time = next_update_time +
                          (next_alarm_sec - 1) * NANOSECONDS_PER_SECOND;
+    if (s->next_alarm_time != next_alarm_time) {
+        s->next_alarm_time = next_alarm_time;
+        qapi_event_send_rtc_alarm(s->next_alarm_time);
+    }

     /* If update_in_progress latched the UIP bit, we must keep the timer
      * programmed to the next second, so that UIP is cleared.  Otherwise,
diff --git a/qapi/misc-target.json b/qapi/misc-target.json
index a00fd821eb..a92959d436 100644
--- a/qapi/misc-target.json
+++ b/qapi/misc-target.json
@@ -24,6 +24,24 @@
   'data': { 'offset': 'int' },
   'if': 'defined(TARGET_ALPHA) || defined(TARGET_ARM) || defined(TARGET_HPPA) || defined(TARGET_I386) || defined(TARGET_MIPS) || defined(TARGET_MIPS64) || defined(TARGET_MOXIE) || defined(TARGET_PPC) || defined(TARGET_PPC64) || defined(TARGET_S390X) || defined(TARGET_SH4) || defined(TARGET_SPARC)' }

+##
+# @RTC_ALARM:
+#
+# Emitted when the guest RTC alarm expire time changes.
+#
+# @expire: expire time of guest RTC alarm, measured in host realtime
+#
+# Example:
+#
+# <-   { "event": "RTC_ALARM",
+#        "data": { "expire": 1593499713567502000 },
+#        "timestamp": { "seconds": 1593413313, "microseconds": 568267 } }
+#
+##
+{ 'event': 'RTC_ALARM',
+  'data': { 'expire': 'int' },
+  'if': 'defined(TARGET_I386)' }
+
 ##
 # @rtc-reset-reinjection:
 #
-- 
2.25.1

